crates/polars-core/src/series/arithmetic/list_borrowed.rs

```diff
@@ -1,177 +1,1026 @@
+                (a, b, lhs.clone())
+            },
+            {
+                let (a, b) = rhs.list_offsets_and_validities_recursive();
+                (a, b, rhs.clone())
+            },
+            lhs.rechunk_validity(),
+            rhs.rechunk_validity(),
+        )? {
+            Either::Left(v) => v,
+            Either::Right(ca) => return Ok(ca.into_series()),
+        };
+
+        Ok(binary_op_exec.finish()?.into_series())
     }
-}
 
-impl ListChunked {
-    /// Helper function for NumOpsDispatchInner implementation for ListChunked.
-    ///
-    /// Run the given `op` on `self` and `rhs`.
-    fn arithm_helper(
+    /// For operations that perform divisions on integers, sets the validity to NULL on rows where
+    /// the denominator is 0.
+    fn prepare_numeric_op_side_validities<T: PolarsNumericType>(
         &self,
-        rhs: &Series,
-        op: &dyn Fn(&Series, &Series) -> PolarsResult<Series>,
-        has_nulls: Option<bool>,
-    ) -> PolarsResult<Series> {
-        polars_ensure!(
-            self.len() == rhs.len(),
-            InvalidOperation: "can only do arithmetic operations on Series of the same size; got {} and {}",
-            self.len(),
-            rhs.len()
-        );
-
-        let mut has_nulls = has_nulls.unwrap_or(false);
-        if !has_nulls {
-            for chunk in self.chunks().iter() {
-                if does_list_have_nulls(chunk) {
-                    has_nulls = true;
-                    break;
-                }
+        lhs: &mut PrimitiveArray<T::Native>,
+        rhs: &mut PrimitiveArray<T::Native>,
+        swapped: bool,
+    ) where
+        PrimitiveArray<T::Native>: polars_compute::comparisons::TotalEqKernel<Scalar = T::Native>,
+        T::Native: Zero + IsFloat,
+    {
+        if !T::Native::is_float() {
+            match self {
+                Self::Div | Self::Rem | Self::FloorDiv => {
+                    let target = if swapped { lhs } else { rhs };
+                    let ne_0 = target.tot_ne_kernel_broadcast(&T::Native::zero());
+                    let validity = combine_validities_and(target.validity(), Some(&ne_0));
+                    target.set_validity(validity);
+                },
+                _ => {},
             }
         }
-        if !has_nulls {
-            for chunk in rhs.chunks().iter() {
-                if does_list_have_nulls(chunk) {
-                    has_nulls = true;
-                    break;
+    }
+
+    /// For list<->primitive where the primitive is broadcasted, we can dispatch to
+    /// `ArithmeticKernel`, which can have optimized codepaths for when one side is
+    /// a scalar.
+    fn apply_array_to_scalar<T: PolarsNumericType>(
+        &self,
+        arr_lhs: PrimitiveArray<T::Native>,
+        r: T::Native,
+        swapped: bool,
+    ) -> PrimitiveArray<T::Native> {
+        match self {
+            Self::Add => ArithmeticKernel::wrapping_add_scalar(arr_lhs, r),
```

Comment:
Do we need to ensure floating point ordering here? Like if `l + r != r + l` for FP operations, should we keep that into account?

Reply:
I think that's a good point - I could add it to the list kernels given that we have the access to `swapped` flag, but currently we don't provide float summation ordering guarantees even in our primitive numeric kernels -

https://github.com/pola-rs/polars/blob/194b31ea6a47a103f4401554755ed904a7f2838e/crates/polars-core/src/chunked_array/arithmetic/numeric.rs#L108

I can maybe open an issue afterwards for addressing this?

Reply:
So no action here is required right?

Reply:
All good.